File: Data Structures/Binary Search Tree/BST Conversion/Conversion.java
import java.util.*;

class Conversion {
    static class Node {
        int data;
        Node left, right;
        Node(int value) {
            data = value;
            left = right = null;
        }
    }

    static void inorderTraversal(Node root, List<Integer> values) {
        if (root == null) {
            return;
        }
        inorderTraversal(root.left, values);
        values.add(root.data);
        inorderTraversal(root.right, values);
    }

    static void assignInorder(Node root, List<Integer> values, int[] index) {
        if (root == null) {
            return;
        }
        assignInorder(root.left, values, index);
        root.data = values.get(index[0]);
        index[0]++;
        assignInorder(root.right, values, index);
    }

    static Node buildTree() {
        try (Scanner sc = new Scanner(System.in)) {
            System.out.print("Enter the root node value: ");
            int value = sc.nextInt();
            Node root = new Node(value);

            Queue<Node> q = new LinkedList<>();
            q.add(root);

            while (!q.isEmpty()) {
                Node current = q.poll();

                System.out.print("Enter left child of " + current.data + " (-1 for no child): ");
                value = sc.nextInt();
                if (value != -1) {
                    current.left = new Node(value);
                    q.add(current.left);
                }

                System.out.print("Enter right child of " + current.data + " (-1 for no child): ");
                value = sc.nextInt();
                if (value != -1) {
                    current.right = new Node(value);
                    q.add(current.right);
                }
            }

            return root;
        }
    }

    static void convert(Node root) {
        List<Integer> values = new ArrayList<>();
        inorderTraversal(root, values);
        Collections.sort(values);

        int[] index = {0};
        assignInorder(root, values, index);
    }

    static void inorderPrint(Node root) {
        if (root == null) {
            return;
        }
        inorderPrint(root.left);
        System.out.print(root.data + " ");
        inorderPrint(root.right);
    }

    public static void main(String[] args) {
        Node root = buildTree();

        System.out.print("In-order traversal of original tree: ");
        inorderPrint(root);
        System.out.println();

        convert(root);

        System.out.print("In-order traversal of the converted BST: ");
        inorderPrint(root);
        System.out.println();
    }
}

File: Data Structures/Binary Search Tree/BST Conversion/Conversion.py
class Node:
    def __init__(self, value):
        self.data = value
        self.left = None
        self.right = None

def inorderTraversal(root, values):
    if root is None:
        return
    inorderTraversal(root.left, values)
    values.append(root.data)
    inorderTraversal(root.right, values)

def assignInorder(root, values, index):
    if root is None:
        return
    assignInorder(root.left, values, index)
    root.data = values[index[0]]
    index[0] += 1
    assignInorder(root.right, values, index)

def build():
    value = int(input("Enter the root node value: "))
    root = Node(value)

    queue = [root]
    while queue:
        current = queue.pop(0)

        value = int(input(f"Enter left child of {current.data} (-1 for no child): "))
        if value != -1:
            current.left = Node(value)
            queue.append(current.left)

        value = int(input(f"Enter right child of {current.data} (-1 for no child): "))
        if value != -1:
            current.right = Node(value)
            queue.append(current.right)

    return root

def convert(root):
    values = []
    inorderTraversal(root, values)
    values.sort()

    index = [0]
    assignInorder(root, values, index)

def ioPrint(root):
    if root is None:
        return
    ioPrint(root.left)
    print(root.data, end=" ")
    ioPrint(root.right)

if __name__ == "__main__":
    root = build()

    print("In-order traversal of original tree: ", end="")
    ioPrint(root)
    print()

    convert(root)

    print("In-order traversal of the converted BST: ", end="")
    ioPrint(root)
    print()
